src/indexable/dataloader/mod.rs

```diff
@@ -52,13 +58,15 @@ where
 
 impl<D, S, C> DataLoader<D, S, C>
 where
-    D: Dataset + Sync,
-    S: Sampler,
-    C: Collate<D::Sample>,
+    D: Dataset + Sync + Send + 'static,
+    S: Sampler + Send + 'static,
+    C: Collate<D::Sample> + Send + 'static,
     D::Sample: Send,
+    C::Output: Send,
 {
-    /// Return not owning iterator over the dataloader.
-    pub fn iter(&self) -> SingleProcessDataLoaderIter<'_, D, S, C> {
+    /// Return owning iterator over the dataloader.
+    /// TODO: Find a way to not consume the Dataloader
+    pub fn iter(self) -> SingleProcessDataLoaderIter<D, S, C> {
```

Comment:
DataLoaderBuilder can return an `Arc<DataLoader>` instead of `DataLoader` this way `SingleProcessDataLoaderIter` will not consume the `Dataloader` (moved in a thread)

Reply:
See next commit where dataset is stored as Arc<Dataset> inside Dataloader

Reply:
Yeah, I've also come to an `Arc` when iterating during the designing of the API. At the end, I choose to mimic what the standard iterator do: 
- `Dataloader::into_iter` take owensership
- `Dataloader::iter`  and `Dataloader::iter_mut` don't

I thought it would be less surprising for the end user and also have a lower overhead.
Is the `Arc` necessary to your change or you can go with `iter` if you don' t want to consume the dataset?

Reply:
Unfortunately in order to not consume the Dataloader like here : 
`let mut iter = dataloader.iter();` and be able to iterator over again I didn't find a solution without using `Arc` because of the thread::spawn that move variables (dataloader, sampler ..)

Reply:
Another solution could be to remove `iter` for the `Dataloader` and keep only `into_iter` then we can move Dataset & Sampler into the spawned thread without using an `Arc` but it seems too restrictive WDYT ?

Reply:
FYI @Tudyx I have [two commit](https://github.com/Tudyx/ai-dataloader/pull/17/commits) in this branch now: 
- One with Arc
- One without -> but we must consume the Dataloader in order to create the iterator